File: app/src/main/java/com/schuetz/rust_android_ios/MainActivity.java
package com.schuetz.rust_android_ios;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        try {
            System.loadLibrary("mobcore");
        } catch (UnsatisfiedLinkError e) {
            Log.e("MainActivity", "Load library error: " + e);
            return;
        }

        final MyRustStruct myRustStruct = new MyRustStruct();

        final TextView greetingTextView = findViewById(R.id.greetingLabel);
        greetingTextView.setText(myRustStruct.greet("Ivan"));

        final TextView numberTextView = findViewById(R.id.numberLabel);
        final int res = myRustStruct.add(2);
        numberTextView.setText(getString(R.string.rust_add_text, res));

        final TextView callbackTextView = findViewById(R.id.callbackLabel);
        myRustStruct.function_with_callback(new Callback() {
            @Override
            public void call(final int a_number, final boolean a_boolean) {
            callbackTextView.setText("Got callback result: a_number: " + a_number + ", a_boolean: " + a_boolean);
            }
        });

        final TextView eventsTextView = findViewById(R.id.eventsLabel);
        myRustStruct.observe(new Callback() {
            @Override
            public void call(final int a_number, final boolean a_boolean) {
                eventsTextView.post(new Runnable() {
                    @Override
                    public void run() {
                        eventsTextView.setText("Received event: a_number: " + a_number + ", a_boolean: " + a_boolean);
                    }
                });
            }
        });
        myRustStruct.send_to_observers(1);
        eventsTextView.postDelayed(new Runnable() {
            @Override
            public void run() {
                myRustStruct.send_to_observers(2);
            }
        }, 2000);
    }
}

File: app/src/main/java/com/schuetz/rust_android_ios/MainActivity.kt
package com.schuetz.rust_android_ios

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.util.Log
import android.widget.TextView

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        try {
            System.loadLibrary("mobcore")
        } catch (e: UnsatisfiedLinkError) {
            Log.e("MainActivity", "Load library error: $e")
            return
        }

        val myRustStruct = MyRustStruct()

        val greetingTextView = findViewById<TextView>(R.id.greetingLabel)
        greetingTextView.text = myRustStruct.greet("Ivan")

        val numberTextView = findViewById<TextView>(R.id.numberLabel)
        val res = myRustStruct.add(2)
        numberTextView.text = getString(R.string.rust_add_text, res)

        val callbackTextView = findViewById<TextView>(R.id.callbackLabel)
        myRustStruct.function_with_callback { a_number, a_boolean ->
            callbackTextView.text = "Got callback result: a_number: $a_number, a_boolean: $a_boolean"
        }

        val eventsTextView = findViewById<TextView>(R.id.eventsLabel)
        myRustStruct.observe { a_number, a_boolean ->
            eventsTextView.post {
                eventsTextView.text = "Received event: a_number: $a_number, a_boolean: $a_boolean"
            }
        }
        myRustStruct.send_to_observers(1)
        eventsTextView.postDelayed({ myRustStruct.send_to_observers(2) }, 2000)
    }
}
